File: other/Solution309.java
class Solution309 {
    public int maxProfit(int[] prices) {
        if(prices.length == 0){
            return 0;
        }
    	int days = prices.length;
    	int[][] dp = new int[days][3];
    	dp[0][0] = -prices[0];
    	dp[0][1] = 0; dp[0][2] = 0;
    	for(int i = 1; i < days; i++){
    		dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][2] - prices[i]);
    		dp[i][1] = dp[i - 1][0] + prices[i];
    		dp[i][2] = Math.max(dp[i - 1][2], dp[i - 1][1]);
    	}
    	return Math.max(dp[days - 1][1], dp[days - 1][2]);
    }
}

File: C++/Solution309.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
#include<queue>
using namespace std;

class Solution309 {
public:
    int maxProfit(vector<int>& prices) {
        int size = prices.size();
        int x = -prices[0], y = 0, z = 0;
        for(int i = 1; i < size; i++){
            int curx = max(x, z - prices[i]);
            int cury = x + prices[i];
            int curz = max(y, z);
            x = curx, y = cury, z = curz;
        }
        return max(y, z);
    }
};

void swap(int& x, int& y){
    x += y;
    y = x - y;
    x -= y;
}

int main(){
    int x = 5, y= 3;
    swap(x, y);
    cout << x << " " << y << endl;
    return 0;
}